src/bvals/bvals_swarm.cpp

```diff
@@ -0,0 +1,162 @@
+//========================================================================================
+// Athena++ astrophysical MHD code
+// Copyright(C) 2014 James M. Stone <jmstone@princeton.edu> and other code contributors
+// Licensed under the 3-clause BSD License, see LICENSE file for details
+//========================================================================================
+// (C) (or copyright) 2020-2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+
+//! \file bvals_swarm.cpp
+//  \brief functions that apply BCs for SWARMs
+
+#include "bvals/bvals_interfaces.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+
+#include "basic_types.hpp"
+#include "globals.hpp"
+#include "mesh/mesh.hpp"
+#include "parameter_input.hpp"
+#include "utils/buffer_utils.hpp"
+#include "utils/error_checking.hpp"
+
+namespace parthenon {
+
+BoundarySwarm::BoundarySwarm(std::weak_ptr<MeshBlock> pmb)
+    : bswarm_index(), pmy_block(pmb), pmy_mesh_(pmb.lock()->pmy_mesh) {
+#ifdef MPI_PARALLEL
+  swarm_id_ = pmb.lock()->pbval->bvars_next_phys_id_;
+#endif
+
+  InitBoundaryData(bd_var_);
+}
+
+void BoundarySwarm::InitBoundaryData(BoundaryData<> &bd) {
+  auto pmb = GetBlockPointer();
+  NeighborIndexes *ni = pmb->pbval->ni;
+
+  bd.nbmax = pmb->pbval->maxneighbor_;
+
+  for (int n = 0; n < bd.nbmax; n++) {
+    bd.flag[n] = BoundaryStatus::waiting;
+#ifdef MPI_PARALLEL
+    bd.req_send[n] = MPI_REQUEST_NULL;
+    bd.req_recv[n] = MPI_REQUEST_NULL;
+#endif
+  }
+}
+
+void BoundarySwarm::SetupPersistentMPI() {
+#ifdef MPI_PARALLEL
+  std::shared_ptr<MeshBlock> pmb = GetBlockPointer();
+  int &mylevel = pmb->loc.level;
```

Comment:
mylevel is not used?

Reply:
Yes right thanks. These were needed for the `Variable` code because how you pack depends on your relative AMR level but that isn't an issue for particles.